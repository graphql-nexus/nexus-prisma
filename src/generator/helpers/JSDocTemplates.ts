import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import { repeat } from 'lodash'

export function jsDocForModel(model: DMMF.Model): string {
  return model.documentation
    ? endent`
        /**
         * ${model.documentation}
         */
      `
    : missingModelDocumentation(model)
}

export function jsDocForField({ field, model }: { field: DMMF.Field; model: DMMF.Model }): string {
  return field.documentation
    ? endent`
        /**
         * ${field.documentation}
         */
      `
    : missingFieldDocumentation({ field, model })
}

export function missingModelDocumentation({ name }: DMMF.Model): string {
  return endent`
    /**
     * ### ðŸ“” Missing Model Documentation for \`${name}\`
     * 
     * Get JSDoc documentation for this model automatically by documenting
     * it in your Prisma Schema âœ¨!
     * 
     * To document a model add a tripple slash comment above it.
     * 
     * Learn more about Prisma Schema comments [here](https://www.prisma.io/docs/concepts/components/prisma-schema#comments).
     * 
     * ${jsDocDiviion('Example 1')}
     * 
     * \`\`\`prisma
     * /// Lorem ipsum dolor sit amet...
     * model ${name} {
     *   id  String  @id
     * }
     * \`\`\`
     * 
     * ${jsDocDiviion('Example 2')}
     * 
     * \`\`\`prisma
     * /// A user record. A user maps 1:1 with a person. Not
     * /// to be confused with an \`Account\` which on person
     * /// may have multiple of, all linked to a single \`User\`.
     * model User {
     *   id  String  @id
     * }
     * \`\`\`
     */
  `
}

export function missingFieldDocumentation({
  field,
  model,
}: {
  field: DMMF.Field
  model: DMMF.Model
}): string {
  return endent`
    /**
     * ### ðŸ“” Missing Field Documentation for \`${field.name}\`
     * 
     * Get JSDoc documentation for this field automatically by documenting
     * it in your Prisma Schema âœ¨!
     * 
     * To document a field add a tripple slash comment above it.
     * 
     * Learn more about Prisma Schema comments [here](https://www.prisma.io/docs/concepts/components/prisma-schema#comments).
     * 
     * ${jsDocDiviion('Example 1')}
     * 
     * \`\`\`prisma
     * model ${model.name} {
     *   /// Lorem ipsum dolor sit amet.
     *   ${field.name}  ${field.type}${field.isRequired ? '' : '?'}
     * }
     * \`\`\`
     * 
     * ${jsDocDiviion('Example 2')}
     * 
     * \`\`\`prisma
     * model User {
     *   /// Identifier for a user. Will never change once set.
     *   /// More stable than the \`handle\` field  which user
     *   /// _can_ change, but is not human-friendly. Prefer this
     *   /// for machine consumers but prefer \`handle\` for human
     *   /// visible things like URL slugs.
     *   id  String  @id
     * }
     * \`\`\`
     */
  `
}

/**
 * Render a centered title with bars on either side filling left over space.
 *
 * Sized according to standard JSDoc window size in VSCode.
 */
function jsDocDiviion(title: string): string {
  const titlePadding = ' '
  const barLength = Math.floor(MAX_JS_DOC_LENGTH / 2 - title.length - titlePadding.length * 2)
  const bar = repeat('â€“', barLength)
  return endent`
    #### ${bar}${titlePadding}${title}${titlePadding}${bar}
  `
}

const MAX_JS_DOC_LENGTH = 72
