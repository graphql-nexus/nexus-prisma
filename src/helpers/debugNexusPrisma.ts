import debug from 'debug'

const debugName = 'nexus-prisma'

export const d = debug(debugName)

d.log = console.log.bind(console)

export function debugNexusPrisma(...names: string[]) {
  const d2 = debug([debugName].concat(names).join(':'))
  d2.log = console.log.bind(console)
  return d2
}

export function packageRun() {}
