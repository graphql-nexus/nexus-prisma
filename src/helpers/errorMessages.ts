export const GITHUB_NEW_ISSUE_LINK = `https://github.com/prisma/nexus-prisma/issues/new/choose`

export const GITHUB_NEW_DISCUSSION_LINK = `https://github.com/prisma/nexus-prisma/discussions/new`
