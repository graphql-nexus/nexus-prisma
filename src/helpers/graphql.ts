export type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

export const graphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
  Float: 'Float',
  Boolean: 'Boolean',
  Int: 'Int',
} as const
