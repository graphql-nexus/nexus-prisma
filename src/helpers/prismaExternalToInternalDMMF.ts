// @ts-expect-error Private api
import * as PrismaClientGenerator from '@prisma/client/generator-build'
import { DMMF, DMMF as ExternalDMMF } from '@prisma/generator-helper'

export const externalToInternalDmmf = PrismaClientGenerator.externalToInternalDmmf as (
  document: ExternalDMMF.Document
) => DMMF.Document
