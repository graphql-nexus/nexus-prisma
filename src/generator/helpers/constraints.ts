import { DMMF } from '@prisma/client/runtime'
import { RecordUnknown } from '../../helpers/utils'

/**
 * Find the unique identifiers necessary to indentify a field
 *
 * Unique fields for a model can be one of (in this order):
 * 1. One (and only one) field with an @id annotation
 * 2. Multiple fields with @@id clause
 * 3. One (and only one) field with a @unique annotation (if there are multiple, use the first one)
 * 4. Multiple fields with a @@unique clause
 */
export function resolveUniqueIdentifiers(model: DMMF.Model): string[] {
  // Try finding 1.
  const singleIdField = model.fields.find((f) => f.isId)

  if (singleIdField) {
    return [singleIdField.name]
  }

  // Try finding 2.
  if (model.idFields && model.idFields.length > 0) {
    return model.idFields
  }

  // Try finding 3.
  const singleUniqueField = model.fields.find((f) => f.isUnique)

  if (singleUniqueField) {
    return [singleUniqueField.name]
  }

  // Try finding 4.
  if (model.uniqueFields && model.uniqueFields.length > 0) {
    return model.uniqueFields[0] as string[] // I don't know why typescript want a cast here
  }

  throw new Error(`Unable to resolve a unique identifier for the Prisma model: ${model.name}`)
}

export function findMissingUniqueIdentifiers(
  data: RecordUnknown,
  uniqueIdentifiers: string[]
): string[] | null {
  const missingIdentifiers: string[] = []

  for (const identifier of uniqueIdentifiers) {
    if (!data[identifier]) {
      missingIdentifiers.push(identifier)
    }
  }

  if (missingIdentifiers.length > 0) {
    return missingIdentifiers
  }

  return null
}

export function buildWhereUniqueInput(data: RecordUnknown, uniqueIdentifiers: string[]): RecordUnknown {
  if (uniqueIdentifiers.length === 1) {
    return pickFromRecord(data, uniqueIdentifiers)
  }

  const compoundName = uniqueIdentifiers.join('_')

  return {
    [compoundName]: pickFromRecord(data, uniqueIdentifiers),
  }
}

function pickFromRecord(record: RecordUnknown, keys: string[]) {
  const output: Record<string, unknown> = {}

  for (const identifier of keys) {
    if (record[identifier]) {
      output[identifier] = record[identifier]
    }
  }

  return output
}
