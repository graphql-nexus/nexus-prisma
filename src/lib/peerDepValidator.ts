import dedent from 'dindist'
import * as Semver from 'semver'
import { PackageJson } from 'type-fest'
import { d } from '../helpers/debugNexusPrisma'
import { isModuleNotFoundError } from '../helpers/utils'
import { detectProjectPackageManager, renderAddDeps } from './packageManager'
import kleur = require('kleur')

type Failure =
  // todo
  // | { message: string; kind: 'peer_dep_import_error' }
  // | { message: string; kind: 'peer_dep_missing_from_node_modules' }
  // | { message: string; kind: 'peer_dep_missing_from_package_json' }
  | { message: string; kind: 'peer_dep_not_installed' } // replace with peer_dep_missing_from_package_json
  | { message: string; kind: 'peer_dep_invalid_version' }
  | { message: string; kind: 'peer_dep_package_json_invalid' }
  | { message: string; kind: 'peer_dep_package_json_read_error'; error: unknown }
  | { message: string; kind: 'unexpected_error'; error: unknown }

export const envarSpecs = {
  NO_PEER_DEPENDENCY_CHECK: {
    name: `NO_PEER_DEPENDENCY_CHECK`,
    values: ['true', '1'],
  },
  PEER_DEPENDENCY_CHECK: {
    name: `PEER_DEPENDENCY_CHECK`,
    values: ['false', '0'],
  },
}

export function enforceValidPeerDependencies({ packageJson }: { packageJson: PackageJson }): void {
  if (
    envarSpecs.NO_PEER_DEPENDENCY_CHECK.values.includes(
      process.env[envarSpecs.NO_PEER_DEPENDENCY_CHECK.name] ?? ''
    )
  )
    return

  if (
    envarSpecs.PEER_DEPENDENCY_CHECK.values.includes(process.env[envarSpecs.PEER_DEPENDENCY_CHECK.name] ?? '')
  )
    return

  d('validating peer dependencies')

  const failure = validatePeerDependencies({ packageJson })

  if (failure) {
    console.error(failure.message)

    if ('error' in failure) {
      console.error(failure.error)
    }

    if (failure.kind === 'peer_dep_not_installed') {
      process.exit(1)
    }
  }
}

/**
 * Check that the given package's peer dependency requirements are met.
 *
 * When envar skipping enabled then:
 *
 * 1. NO-op if PEER_DEPENDENCY_CHECK envar is set to false or 0
 * 2. NO-op if NO_PEER_DEPENDENCY_CHECK envar is set to true or 1
 */
export function validatePeerDependencies({ packageJson }: { packageJson: PackageJson }): null | Failure {
  try {
    const peerDependencies = packageJson['peerDependencies'] ?? []

    for (const [pdName, _] of Object.entries(peerDependencies)) {
      const failure = validatePeerDependencyRangeSatisfied({
        peerDependencyName: pdName,
        requireer: packageJson,
      })

      if (failure) return failure
    }
  } catch (error: unknown) {
    const code = `unexpected_error`
    return {
      kind: code,
      message: renderWarning({
        title: `Something went wrong while trying to validate peer dependencies`,
        code,
        reason: error instanceof Error ? error.message : String(error),
        consequence: `There seems to be a bug so the regular correctness checks of the peer dep checker cannot be carried out now. You are on your own.`,
        solution: `Please report this issue.`,
      }),
      error,
    }
  }

  return null
}

export function validatePeerDependencyRangeSatisfied({
  peerDependencyName,
  requireer,
}: {
  peerDependencyName: string
  requireer: PackageJson
}): null | Failure {
  let pdPackageJson: PackageJson

  try {
    // Use the Node dep lookup algorithm
    // eslint-disable-next-line
    pdPackageJson = require(`${peerDependencyName}/package.json`) as PackageJson
  } catch (error: unknown) {
    if (!isModuleNotFoundError(error)) {
      return {
        kind: 'peer_dep_package_json_read_error',
        message: `Peer dependency check found ${peerDependencyName} requried by ${
          requireer.name ?? '<undefined>'
        } to be installed but encountered an error while reading its package.json.`,
        error,
      }
    }

    const code = 'peer_dep_not_installed'
    return {
      kind: code,
      message: renderError({
        title: `Peer dependency validation check failed.`,
        // prettier-ignore
        reason: dedent`${kleur.green(peerDependencyName)} is a peer dependency required by ${renderPackageJsonField(requireer,'name')}. But you have not installed it into this project yet.`,
        code,
        // prettier-ignore
        solution: `Please run \`${kleur.green(renderAddDeps(detectProjectPackageManager(),[peerDependencyName]))}\`.`,
        consequence: `Your project may not work correctly.`,
      }),
    }
  }

  const pdVersion = pdPackageJson.version
  const pdVersionRangeSupported = requireer.peerDependencies?.[peerDependencyName]

  // npm enforces that package manifests have a valid "version" field so this
  // case _should_ never happen under normal circumstances.
  if (!pdVersion) {
    const code = 'peer_dep_package_json_invalid'
    return {
      kind: code,
      message: renderWarning({
        title: `Peer dependency validation check failed unexpectedly.`,
        // prettier-ignore
        reason: `${renderPackageJsonField(requireer, 'name')} requires peer dependency ${renderPackageJsonField(pdPackageJson, 'name')}. No version info for ${renderPackageJsonField(pdPackageJson, 'name')} could be found in its package.json thus preventing a check if its version satisfies the peer dependency version range.`,
        consequence: `Peer dep validator checks cannot be carried out so you are on your own.`,
        code,
      }),
    }
  }

  if (!pdVersionRangeSupported) {
    const code = `unknown`
    console.warn(
      renderWarning({
        title: `Peer dependency validation check failed unexpectedly.`,
        // prettier-ignore
        reason: `${renderPackageJsonField(requireer, 'name')} apparently requires peer dependency ${renderPackageJsonField(pdPackageJson, 'name')} yet ${renderPackageJsonField(pdPackageJson, 'name')} is not listed in the peer dependency listing of ${renderPackageJsonField(requireer, 'name')}.`,
        consequence: `There seems to be a bug so the regular correctness checks of the peer dep checker cannot be carried out now. You are on your own. Please report this issue.`,
        code,
      })
    )
    return null
  }

  if (Semver.satisfies(pdVersion, pdVersionRangeSupported)) {
    return null
  }

  return {
    kind: 'peer_dep_invalid_version',
    message: renderWarning({
      title: `Peer dependency validation check failed`,
      // prettier-ignore
      reason: `${renderPackageJsonField(requireer, 'name')}@${renderPackageJsonField(requireer, 'version')} does not officially support ${renderPackageJsonField(pdPackageJson, 'name')}@${renderPackageJsonField(pdPackageJson, 'version')}. The officially supported range is: \`${pdVersionRangeSupported}\`.`,
      consequence: `This could lead to undefined behaviors and bugs.`,
      code: `peer_dep_invalid_version`,
    }),
  }
}

//prettier-ignore
const prettyPrintedDisableGuide = dedent`
  HOW TO DISABLE:
  
    You can disable this peer dependency check by setting one of two environment variables. Their specs are:

      ${envarSpecs.NO_PEER_DEPENDENCY_CHECK.name} = ${envarSpecs.NO_PEER_DEPENDENCY_CHECK.values.map(_=>`'${_}'`).join(` | `)}
      ${envarSpecs.PEER_DEPENDENCY_CHECK.name}    = ${envarSpecs.PEER_DEPENDENCY_CHECK.values.map(_=>`'${_}'`).join(` | `)}

    Examples:

      NO_PEER_DEPENDENCY_CHECK='true'
      NO_PEER_DEPENDENCY_CHECK='1'
      PEER_DEPENDENCY_CHECK='false'
      PEER_DEPENDENCY_CHECK='0'
`

type DiagnosticInfo = {
  title: string
  code: string
  reason: string
  consequence: string
  solution?: string
}

function renderError(params: DiagnosticInfo): string {
  const solution = params.solution ? `\n\nSOLUTION: ${params.solution}` : ''
  // prettier-ignore
  return `${kleur.red('ERROR:')} ${params.title}\n\nREASON: ${params.reason}\n\nCONSEQUENCE: ${params.consequence}${solution}\n\n${prettyPrintedDisableGuide}\n\nCODE: ${params.code}`
}

function renderWarning(params: DiagnosticInfo): string {
  const solution = params.solution ? `\n\nSOLUTION: ${params.solution}` : ''
  // prettier-ignore
  return `${kleur.yellow('WARNING:')} ${params.title}\n\nREASON: ${params.reason}\n\nCONSEQUENCE: ${params.consequence}${solution}\n\n${prettyPrintedDisableGuide}\n\nCODE: ${params.code}`
}

function renderPackageJsonField(packageJson: PackageJson, fieldName: keyof PackageJson): string {
  return kleur.yellow(
    packageJson[fieldName] === undefined ? `<${fieldName} is undefined>` : String(packageJson[fieldName])
  )
}
