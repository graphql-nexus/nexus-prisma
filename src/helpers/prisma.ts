import { DMMF } from '@prisma/client/runtime'
import dedent from 'dindist'
import kleur from 'kleur'
import ono from 'ono'
import { inspect } from 'util'
import { detectProjectPackageManager, renderRunBin } from '../lib/packageManager'
import { d } from './debugNexusPrisma'
import { GITHUB_NEW_DISCUSSION_LINK } from './errorMessages'

/**
 * Given a package loader, attempt to get the Prisma Client DMMF.
 *
 * @remarks Only the given require function is truly import, the rest is used for better error messages.
 *
 *          This function is designed to support working with bundlers. Specifically `ncc` has been
 *          tested.
 *
 *          This function intentionally does not do the `require`/`import` itself, leaving that to
 *          upstream code to handle in static ways that bundlers will be able to process.
 */
export const getPrismaClientDmmf = (packageLoader: {
  /**
   * A function that must return the Prisma Client Package
   */
  require: () => unknown
  /**
   * The import specifier being used (the from "..." part)
   */
  importId: string
  /**
   * The resolved import specifier being used. This can be different than important ID in two ways:
   *
   * 1. NodeJS lookp algorithm
   * 2. Bundlers that rewrite import paths
   */
  importIdResolved: string
}): DMMF.Document => {
  d('get dmmf from @prisma/client')

  let prismaClientPackage: unknown

  // prettier-ignore
  const printedImportId = `${kleur.yellow(packageLoader.importId)} (resolved as ${packageLoader.importIdResolved})`

  try {
    // eslint-disable-next-line
    prismaClientPackage = packageLoader.require()
  } catch (error) {
    // prettier-ignore
    throw ono(error as Error, dedent`
      Failed to get Prisma Client DMMF. An error occured while trying to import it from ${printedImportId}.
    `)
  }

  if (!(typeof prismaClientPackage === 'object' && prismaClientPackage !== null)) {
    // prettier-ignore
    throw new Error(dedent`
      Failed to get Prisma Client DMMF. It was imported from ${printedImportId} but was not the expected type. Got:

      ${inspect(prismaClientPackage)}
    `)
  }

  const prismaClientPackageObject = prismaClientPackage as Record<string, unknown>

  // eslint-disable-next-line
  if (!prismaClientPackageObject.dmmf) {
    // prettier-ignore
    throw new Error(dedent`
      Failed to get Prisma Client DMMF. It was imported from ${printedImportId} but did not contain "dmmf" data. Got:

      ${inspect(prismaClientPackage)}

      This usually means that you need to run Prisma Client generation. Please run ${renderRunBin(detectProjectPackageManager(), `prisma generate`)}.
      If that does not solve your problem, you can get community help by opening a discussion at ${kleur.yellow(GITHUB_NEW_DISCUSSION_LINK)}.
    `)
  }

  // Simple duck type to sanity check we got good data at runtime.

  const dmmf = prismaClientPackageObject.dmmf as DMMF.Document

  const expectedFields = ['datamodel', 'schema', 'mappings'] as const

  if (expectedFields.find((fieldName) => dmmf[fieldName] && typeof dmmf[fieldName] !== 'object')) {
    throw new Error(dedent`
      The DMMF imported from ${packageLoader.importId} appears to be invalid. Missing one/all of expected fields: 
    `)
  }

  return dmmf
}

export type PrismaScalarType =
  | 'String'
  | 'Boolean'
  | 'Int'
  | 'BigInt'
  | 'Float'
  | 'Decimal'
  | 'DateTime'
  | 'Json'
  | 'Bytes'
