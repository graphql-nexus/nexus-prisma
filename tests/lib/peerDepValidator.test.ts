import dedent from 'dindist'
import { kont } from 'kont'
import { Providers } from 'kont/providers'
import { merge, omit } from 'lodash'
import { PackageJson } from 'type-fest'
import { envarSpecs } from '../../src/lib/peerDepValidator'
import { assertBuildPresent } from '../__helpers__/helpers'
import { project } from '../__providers__/project'

/** Setup */

const badJson = ';'

const importer = {
  name: 'alpha',
}

const peerDep = {
  name: 'charlie',
}

const ctx = kont()
  .useBeforeAll(Providers.Dir.create())
  .useBeforeAll(Providers.Run.create())
  .useBeforeAll(project())
  .done()

beforeAll(() => {
  assertBuildPresent()

  // setup alpha dep that has peer dep requirements

  ctx.runOrThrow(`yarn add kleur semver tslib debug fs-jetpack dindist --production`)

  ctx.fs.write(`node_modules/${importer.name}/package.json`, {
    name: importer.name,
    version: '1.0.0',
    main: 'dist/index.js',
  })

  ctx.fs.copy(`${process.cwd()}/dist-cjs`, `${ctx.fs.cwd()}/node_modules/${importer.name}/dist`)

  ctx.fs.write(
    'validatePeerDependencies.js',
    dedent`
      const assert = require('assert')
      const { validatePeerDependencies } = require('${importer.name}/dist/lib/peerDepValidator')

      const packageJson = require('${importer.name}/package.json')
      assert(packageJson)

      const failure = validatePeerDependencies({
        packageJson,
      })

      console.log(failure)
    `
  )

  ctx.fs.write(
    'enforceValidPeerDependencies.js',
    dedent`
      const assert = require('assert')
      const { enforceValidPeerDependencies } = require('${importer.name}/dist/lib/peerDepValidator')

      const packageJson = require('${importer.name}/package.json')
      assert(packageJson)

      enforceValidPeerDependencies({
        packageJson,
      })
    `
  )
})

beforeEach(() => {
  setupPeerDepRequirement({
    name: peerDep.name,
    range: '2.0.x',
    optional: false,
  })

  ctx.fs.remove(`node_modules/${peerDep.name}`)
})

/** Helpers */

function setupDep({
  name,
  version = '0.0.0',
  main = 'exports.hello = "world"',
  packageJson = (x) => x,
}: {
  name: string
  version?: string
  main?: string
  packageJson?: (defaultPackageJson: PackageJson) => PackageJson | string
}): void {
  const depDir = `node_modules/${name}`
  ctx.fs.write(`${depDir}/package.json`, packageJson({ name, version, main: './index.js' }))
  ctx.fs.write(`${depDir}/index.js`, main)
}

function setupPeerDepRequirement({
  name,
  range,
  optional,
}: {
  name: string
  range: string
  optional: boolean
}) {
  const old = ctx.fs.read(`node_modules/${importer.name}/package.json`, 'json')
  ctx.fs.write(
    `node_modules/${importer.name}/package.json`,
    merge(old, {
      peerDependencies: {
        [name]: range,
      },
      peerDependenciesMeta: {
        [name]: { optional },
      },
    })
  )
}

function runValidatePeerDependencies() {
  return ctx.runOrThrow('node validatePeerDependencies.js', {
    env: {
      ...process.env,
      FORCE_COLOR: '0',
    },
  })
}

function runEnforceValidPeerDependencies(params?: { env?: Record<string, string> }) {
  return ctx.runOrThrow('node enforceValidPeerDependencies.js', {
    env: {
      ...process.env,
      FORCE_COLOR: '0',
      ...params?.env,
    },
  })
}

/** Tests */

describe('ValidatePeerDependencies', () => {
  it('if peer dep missing, then returns failure', () => {
    expect(runValidatePeerDependencies().stdout).toMatchSnapshot()
  })

  it('if peer dep is optional, then no check is made against it', () => {
    setupPeerDepRequirement({
      name: peerDep.name,
      range: '2.0.x',
      optional: true,
    })
    expect(runValidatePeerDependencies().stdout).toMatchSnapshot()
  })

  it('if peer dep installed, but upon import JSON parsing fails somehow, then fails', () => {
    setupDep({
      name: peerDep.name,
      packageJson() {
        return badJson
      },
    })
    expect(runValidatePeerDependencies().stdout).toMatch(/peer_dep_package_json_read_error/)
  })

  it('if project peer dep version does not satisfy required range, then returns failure', () => {
    setupDep({
      name: peerDep.name,
      version: '1.0.0',
    })
    expect(runValidatePeerDependencies().stdout).toMatchSnapshot()
  })

  it('if peer dep version satisfies required range, then returns null', () => {
    setupDep({
      name: peerDep.name,
      version: '2.0.1',
    })
    expect(runValidatePeerDependencies().stdout).toMatchSnapshot()
  })

  it('if peer dep package.json missing version field, then returns failure', () => {
    setupDep({
      name: peerDep.name,
      version: '1.0.0',
      packageJson(packageJson) {
        return omit(packageJson, 'version')
      },
    })
    expect(runValidatePeerDependencies().stdout).toMatchSnapshot()
  })
})

describe('enforceValidPeerDependencies', () => {
  it(`if ${[envarSpecs.PEER_DEPENDENCY_CHECK.name]}=false|0 then no validation happens`, () => {
    // prettier-ignore
    expect(runEnforceValidPeerDependencies({ env: { [envarSpecs.PEER_DEPENDENCY_CHECK.name]: 'false' } }).stdout).toEqual(``)
    // prettier-ignore
    expect(runEnforceValidPeerDependencies({ env: { [envarSpecs.PEER_DEPENDENCY_CHECK.name]: '0' } }).stdout).toEqual(``)
  })

  it(`if ${[envarSpecs.NO_PEER_DEPENDENCY_CHECK.name]}=true|1 then no validation happens`, () => {
    // prettier-ignore
    expect(runEnforceValidPeerDependencies({ env: { [envarSpecs.NO_PEER_DEPENDENCY_CHECK.name]: 'true' } }).stdout).toEqual(``)
    // prettier-ignore
    expect(runEnforceValidPeerDependencies({ env: { [envarSpecs.NO_PEER_DEPENDENCY_CHECK.name]: '1' } }).stdout).toEqual(``)
  })

  it('if peer dependency is missing, than logs and process exits 1', () => {
    expect(() => runEnforceValidPeerDependencies()).toThrowErrorMatchingSnapshot()
  })
})
