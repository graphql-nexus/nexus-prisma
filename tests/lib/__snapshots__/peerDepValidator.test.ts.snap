// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidatePeerDependencies if peer dep missing, then returns failure 1`] = `
"{
  kind: 'peer_dep_not_installed',
  message: 'ERROR: charlie is a peer dependency required by alpha. But you have not installed it into this project yet. Please run \`yarn add charlie\`.'
}"
`;

exports[`ValidatePeerDependencies if peer dep package.json missing version field, then returns failure 1`] = `
"{
  kind: 'peer_dep_invalid_package_json',
  message: 'WARNING: Peer dependency validation check failed unexpectedly. alpha requires peer dependency charlie. No version info for charlie could be found in its package.json thus preventing a check if its version satisfies the peer dependency version range.'
}"
`;

exports[`ValidatePeerDependencies if peer dep version satisfies required range, then returns null 1`] = `"null"`;

exports[`ValidatePeerDependencies if project peer dep version does not satisfy required range, then returns failure 1`] = `
"{
  kind: 'peer_dep_invalid_version',
  message: 'WARNING: Peer dependency validation check failed: alpha@1.0.0 does not officially support charlie@1.0.0. The officially supported range is: \`2.0.x\`. This could lead to undefined behaviors and bugs.'
}"
`;

exports[`enforceValidPeerDependencies if peer dependency is missing, than logs and process exits 1 1`] = `
"{
  kind: 'peer_dep_not_installed',
  message: 'ERROR: charlie is a peer dependency required by alpha. But you have not installed it into this project yet. Please run \`yarn add charlie\`.'
}
ERROR: charlie is a peer dependency required by alpha. But you have not installed it into this project yet. Please run \`yarn add charlie\`."
`;
